File: Dart/Behavioral/command.dart
/// Interface with bussiness logic
/// the implementations do the work
abstract interface class Reciever {}

/// A concrete implementation of reciever
/// implements concretes methods
class Light implements Reciever {
  void turnOn() => print('Light on!');
  void turnOff() => print('Light off!');
}

///Interface with the execute method
abstract interface class Command {
  Command({required this.reciever});

  Reciever reciever;

  void execute();
}

/// Concrete implementation of command
/// Recieves an implmentation of reciever
/// and delegates the operation
class TurnOff extends Command {
  TurnOff({required Light light}) : super(reciever: light);

  @override
  void execute() => (reciever as Light).turnOff();
}

/// Concrete implementation of command
/// Recieves an implmentation of reciever
/// and delegates the operation
class TurnOn extends Command {
  TurnOn({required Light light}) : super(reciever: light);

  @override
  void execute() => (reciever as Light).turnOn();
}

/// Inits and delegate the operations
/// to commands.
/// For this case we save a logs history
class Invoker {
  List<String> _history = [];

  void execute({required Command command}) {
    command.execute();
    _history.add('Executed ${command.runtimeType}');
  }

  String getHistory() =>
      _history.fold('', (previusLog, log) => previusLog + '$log\n');
}

void main() {
  final light = Light();
  final invoker = Invoker();

  invoker.execute(command: TurnOn(light: light));
  invoker.execute(command: TurnOff(light: light));
  invoker.execute(command: TurnOn(light: light));

  print('\nLogs:');
  print(invoker.getHistory());

  /// Here is the result:
/*
Light on!
Light off!
Light on!

Logs:
Executed TurnOn
Executed TurnOff
Executed TurnOn
*/
}

File: Swift/Behavioral/command.swift
//
//  command.swift
//
//
//

// Protocol with bussiness logic
// the implementations do the work
protocol Reciever {}

// A concrete implementation of reciever
// implements concretes methods
struct Light: Reciever {
    func turnOn() { print("Light on!") }
    
    func turnOff() { print("Light off!") }
}

// Protocol with the execute method
protocol Command {
    var reciever: any Reciever { get }
    
    func execute()
}

// Concrete implementation of command
// Recieves an implmentation of reciever
// and delegates the operation
struct TurnOff: Command {
    let reciever: any Reciever

    init(light: Light) {
        self.reciever = light
    }

    func execute() {
        (reciever as! Light).turnOff()
    }
}

// Concrete implementation of command
// Recieves an implmentation of reciever
// and delegates the operation
struct TurnOn: Command {
    let reciever: any Reciever

    init(light: Light) {
        self.reciever = light
    }

    func execute() {
        (reciever as! Light).turnOn()
    }
}

// Inits and delegate the operations
// to commands.
// For this case we save a logs history
struct Invoker {
    private var history: [String] = []

    mutating func execute(command: any Command) {
        command.execute()
        history.append("Executed \(type(of: command))")
    }

    func getHistory() -> String {
        history.reduce("") { $0 + "\($1)\n" }
    }
}

// Demo

let light = Light()
var invoker = Invoker()

invoker.execute(command: TurnOn(light: light))
invoker.execute(command: TurnOff(light: light))
invoker.execute(command: TurnOn(light: light))

print("\nLogs:")
print(invoker.getHistory())

// Here is the result:

// Light on!
// Light off!
// Light on!
//
// Logs:
// Executed TurnOn
// Executed TurnOff
// Executed TurnOn
